### PR TITLE
Enzyme: remove closures

### PR DESCRIPTION
**Pull request by wsmoses**
The cons closures should be removed as well, but lets see if this makes things better already

### CODE REVIEW

File: ext/OptimizationEnzymeExt.jl [hunk 1]
```diff
@@ -15,13 +15,47 @@ isdefined(Base, :get_extension) ? (using Enzyme) : (using ..Enzyme)
     end
 end
 
+function inner_grad(θ, bθ, f, p, args::VarArgs{Any, N}) where N
```
Comment (ChrisRackauckas) [inline ext/OptimizationEnzymeExt.jl:18]:
```suggestion
function inner_grad(θ, bθ, f, p, args::Vararg{Any, N}) where N
```

File: ext/OptimizationEnzymeExt.jl [hunk 2]
```diff
@@ -132,10 +132,14 @@ function OptimizationBase.instantiate_function(f::OptimizationFunction{true}, x,
...
                 end
                 Enzyme.autodiff(Enzyme.Forward,
-                    f2,
+                    cons_f2,
                     Enzyme.BatchDuplicated(θ, vdθ),
                     Enzyme.BatchDuplicated(bθ, vdbθ),
-                    Const(fncs[i]))
+                    Const(f.cons),
+                    Const(p),
+                    Const(num_const),
```
Comment (ChrisRackauckas) [inline ext/OptimizationEnzymeExt.jl:140]:
```suggestion
                    Const(num_cons),
```

File: ext/OptimizationEnzymeExt.jl [hunk 3]
```diff
@@ -15,13 +15,47 @@ isdefined(Base, :get_extension) ? (using Enzyme) : (using ..Enzyme)
...
+        firstapply,
+        Active,
+        f,
+        Enzyme.Duplicated(x, dx),
+        Const(p),
+        Const.(args)...)
+    return dx
+end
+
+function inner_cons(x, p, num_cons, i)
```
Comment (ChrisRackauckas) [inline ext/OptimizationEnzymeExt.jl:41]:
```suggestion
function inner_cons(f, x, p, num_cons, i)
```

File: ext/OptimizationEnzymeExt.jl [hunk 4]
```diff
@@ -15,13 +15,47 @@ isdefined(Base, :get_extension) ? (using Enzyme) : (using ..Enzyme)
...
+    res = zeros(eltype(x), num_cons)
+    f.cons(res, x, p)
+    return res[i]
+end
+
+function cons_f2(x, dx, fcons, p, num_cons, i)
+    Enzyme.autodiff_deferred(Enzyme.Reverse, inner_cons, Active, Enzyme.Duplicated(x, dx), Const(p), Const(num_cons), Const(i))
```
Comment (ChrisRackauckas) [inline ext/OptimizationEnzymeExt.jl:48]:
```suggestion
    Enzyme.autodiff_deferred(Enzyme.Reverse, inner_cons, Active, fcons, Enzyme.Duplicated(x, dx), Const(p), Const(num_cons), Const(i))
```

Won't we need to zero and duplicate the function if it's a closure?

**Comment by Vaibhavdixit02** [comment]
Superseded by #60

File: ext/OptimizationEnzymeExt.jl (hunk 4 above)
Comment (Vaibhavdixit02) [reply ext/OptimizationEnzymeExt.jl:48]: I don't follow, can you add more detail what you mean by zeroing the function? It doesn't need to be duplicated it should be Const iiuc (done that in #60)

**Pull request closed without merging**